waiter/src/waiter/token_watch.clj

```diff
@@ -79,72 +82,79 @@
-                              current-state
-                              (let [[index-event next-state]
-                                    (if (some? token-index-entry)
-                                      ; If index-entry retrieved from kv-store exists then treat as UPDATE
-                                      ; (includes token creation and soft deletion)
-                                      [(make-index-event :UPDATE token-index-entry)
-                                       (assoc-in current-state [:token->index token] token-index-entry)]
-                                      ; index-entry doesn't exist then treat as DELETE
-                                      [(make-index-event :DELETE {:token token})
-                                       (assoc current-state :token->index (dissoc token->index token))])
-                                    open-chans (->> [index-event]
-                                                    (make-index-event :EVENTS)
-                                                    (send-event-to-channels! watch-chans))]
-                                (assoc next-state :watch-chans open-chans)))))
+                          (let [{:keys [token x-cid] :as internal-event} msg]
+                            (cid/with-correlation-id
+                              (str "token-watch-maintainer" "." x-cid)
+                              (log/info "token-watch-maintainer received an internal index event" internal-event)
+                              (let [token-index-entry (token/get-token-index kv-store token :refresh true)
+                                    local-token-index-entry (get token->index token)]
+                                (if (= token-index-entry local-token-index-entry)
+                                  ; There is no change detected, so no event to be reported
+                                  current-state
+                                  (let [[index-event next-state]
+                                        (if (some? token-index-entry)
+                                          ; If index-entry retrieved from kv-store exists then treat as UPDATE
+                                          ; (includes token creation and soft deletion)
+                                          [(make-index-event :UPDATE token-index-entry)
+                                           (assoc-in current-state [:token->index token] token-index-entry)]
+                                          ; index-entry doesn't exist then treat as DELETE
+                                          [(make-index-event :DELETE {:token token})
+                                           (assoc current-state :token->index (dissoc token->index token))])
+                                        _ (log/info "token-watch-maintainer sending a token event to watches" {:event index-event})
+                                        open-chans (->> (make-index-event :EVENTS [index-event] :id x-cid)
```

Comment:
CIDs are not expected to be unique, I think we want the events to have their own unique identifier, e.g. by invoking `utils/unique-identifier`. We can link external CIDs to the id via the info log above.